ndonnx/_typed_array/masked_onnx.py

```diff
@@ -0,0 +1,881 @@
+# Copyright (c) QuantCo 2023-2025
+# SPDX-License-Identifier: BSD-3-Clause
+
+from __future__ import annotations
+
+import operator
+from abc import abstractmethod
+from collections.abc import Callable, Sequence
+from types import NotImplementedType
+from typing import TYPE_CHECKING, TypeVar, overload
+
+import numpy as np
+from typing_extensions import Self
+
+from .._dtypes import TY_ARRAY_BASE, DType
+from .._schema import DTypeInfoV1
+from .._types import NestedSequence, OnnxShape, PyScalar
+from . import (
+    TyArrayBase,
+    astyarray,
+    maximum,
+    minimum,
+    onnx,
+    result_type,
+    safe_cast,
+    where,
+)
+
+if TYPE_CHECKING:
+    from spox import Var
+
+    from .indexing import GetitemIndex, SetitemIndex
+    from .onnx import VALUE
+
+
+DTYPE = TypeVar("DTYPE", bound=DType)
+CORE_DTYPES = TypeVar("CORE_DTYPES", bound=onnx.DTypes)
+NCORE_DTYPES = TypeVar("NCORE_DTYPES", bound="NCoreDTypes")
+
+NCORE_NUMERIC_DTYPES = TypeVar("NCORE_NUMERIC_DTYPES", bound="NCoreNumericDTypes")
+NCORE_FLOATING_DTYPES = TypeVar("NCORE_FLOATING_DTYPES", bound="NCoreFloatingDTypes")
+NCORE_INTEGER_DTYPES = TypeVar("NCORE_INTEGER_DTYPES", bound="NCoreIntegerDTypes")
+
+TY_MA_ARRAY_ONNX = TypeVar("TY_MA_ARRAY_ONNX", bound="TyMaArray", covariant=True)
+
+
+class _MaOnnxDType(DType[TY_MA_ARRAY_ONNX]):
+    _unmasked_dtype: onnx._OnnxDType
+
+    def __ndx_create__(
+        self, val: PyScalar | np.ndarray | TyArrayBase | Var | NestedSequence
+    ) -> TY_MA_ARRAY_ONNX:
+        if isinstance(val, np.ma.MaskedArray):
+            data = safe_cast(onnx.TyArray, astyarray(val.data))
+            if val.mask is np.ma.nomask:
+                mask = None
+            else:
+                mask = safe_cast(onnx.TyArrayBool, onnx.const(val.mask))
+            return asncoredata(data, mask).astype(self)
+        else:
+            return asncoredata(self._unmasked_dtype.__ndx_create__(val), None).astype(
+                self
+            )
+
+    def __str__(self) -> str:
+        return type(self).__name__.lower()
+
+    @abstractmethod
```

Comment:
Why is this abstract? Is it not just `self._unmasked_dtype.__ndx_create__(...)` and the same mask logic for all the masked implementations here?

Reply:
We have to specify the array class somewhere where it is accessible at runtime (i.e. not the generic type variable). The `_build` function is usually the place to do so.